### PR TITLE
Feature/dashboard statistics

### PR DESCRIPTION
**Pull request by M52**
User can now see the number of 502/504 errors that recently occurred as well as view the database tables and its size.

### CODE REVIEW

File: src/class-read-logs.php [hunk 1]
```diff
@@ -39,6 +39,26 @@ function get_log_lines( $log, $lines ) {
...
+        }
+
+        $file    = fopen( $log_path, 'r' );
+        $results = array();
+        while (($bufferedline = fgets($file))) {
+            if (preg_match( $regex, $bufferedline )) {
+                array_push( $results, $bufferedline);
+            }
+        }
+        return $results;
```
Comment (fuegas) [inline src/class-read-logs.php]: You're not closing the file you've opened a few lines above.
```suggestion
        fclose( $file );
        return $results;
```

File: src/class-savvii-dashboard.php [hunk 2]
```diff
@@ -162,13 +154,60 @@ function warpdrive_dashboard() {
...
+                                <?php
+                                echo ' '.count( $errors_502 ).' ';
+                                ?>
+                            recent 502 (Bad Gateway) error(s)</p>
+                            <p>WarpDrive detected <?php echo count( $errors_504 ); ?> recent 504 (Gateway Timeout) error(s)</p>
+                        </div>
+                        <div class="pct50">
+                            <?php
+                            $database = new Database();
+                            $results = $database->get_wp_table_sizes(10);
```
Comment (fuegas) [inline src/class-savvii-dashboard.php]: This looks like duplicate code from the database size page, is it an idea to gather the rendering of the data in a function used by both pages?

File: src/class-savvii-dashboard.php [hunk 3]
```diff
@@ -162,13 +154,60 @@ function warpdrive_dashboard() {
...
                     <!-- /Read server logs -->
+                    <!-- Statistics -->
+                    <div class="postbox">
+                        <h2 class="hndle">Statistics</h2>
+                        <div class="inside">
+                            <div class="pct50">
+                            <?php
+                            $logreader = new ReadLogs();
+                            $errors_502 = $logreader->find_match_in_log( 'access', '/HTTP\/(?:1\.1|2)\"\s+502/i');
+                            $errors_504 = $logreader->find_match_in_log( 'access', '/HTTP\/(?:1\.1|2)\"\s+504/i');
```
Comment (fuegas) [inline src/class-savvii-dashboard.php]: You're opening and reading the whole access log twice now, is that what you're after? Or could you give multiple regexes? Or make the regexes a bit smarter and do something with the matched values? For example:
```
/HTTP\/(?:1\.1|2)\"\s+(50[24])/i
```
Then you can use the data from the capture group to run a tally on.

**Comment by M52** [comment]
Reopened so the necessary fixes can be integrated in this PR. Do not attempt to merge yet.

**Comment by M52** [comment]
All issues raised by @fuegas have been adressed.

File: src/class-savvii-dashboard.php [hunk 4]
```diff
@@ -162,13 +155,82 @@ function warpdrive_dashboard() {
...
+                    <!-- Statistics -->
+                    <div class="postbox">
+                        <h2 class="hndle">Statistics</h2>
+                        <div class="inside">
+                            <div class="row">
+                                <?php
+                                $errors = $logreader->find_match_in_log( 'access', '/HTTP\/(?:1\.1|2)\"\s+(50[24])/i');
+                                $errors_502 = 0;
+                                $errors_504 = 0;
+                                foreach ($errors as $error) {
```
Comment (fuegas) [inline src/class-savvii-dashboard.php:195]: Doing business logic in the view isn't really great, the view should only print the values received.

Also, take a look at [`array_count_values`](https://www.php.net/manual/en/function.array-count-values), it should do exactly why you try to do here, but easier.

File: src/class-savvii-dashboard.php [hunk 5]
```diff
@@ -162,13 +155,82 @@ function warpdrive_dashboard() {
...
+                                    if (strpos($error, '502')) {
```
Comment (fuegas) [inline src/class-savvii-dashboard.php:196]: Please read the warning about [return values of strpos](https://www.php.net/manual/en/function.strpos#refsect1-function.strpos-returnvalues) or you might get strange results.

File: src/class-savvii-dashboard.php [hunk 6]
```diff
@@ -162,13 +155,82 @@ function warpdrive_dashboard() {
...
+                                        $errors_502++;
+                                    } elseif (strpos($error, '504')) {
+                                        $errors_504++;
+                                    }
+                                }
+
+                                //$errors_502 = $logreader->find_match_in_log( 'access', '/HTTP\/(?:1\.1|2)\"\s+502/i');
```
Comment (fuegas) [inline src/class-savvii-dashboard.php:203]: Dead code does not belong in production code 😄

File: src/class-savvii-dashboard.php [hunk 7]
```diff
@@ -162,13 +155,82 @@ function warpdrive_dashboard() {
...
+                                <?php
+                                $results = Database::get_wp_table_sizes(10);
+                                ?>
+                                <div class="column">
+                                    <table>
+                                        <tr>
+                                            <th>Table</th>
+                                            <th>Size (MB)</th>
+                                        </tr>
+                                    <?php foreach($results as $row) { ?>
```
Comment (fuegas) [inline src/class-savvii-dashboard.php:219]: As `$results` is quite a vague term for the info you're retrieving and you're only using it once, you can use the function call directly in the `for`.
```suggestion
                                    <?php foreach(Database::get_wp_table_sizes(10) as $row) { ?>
```

File: src/class-savvii-dashboard.php [hunk 8]
```diff
@@ -162,13 +155,82 @@ function warpdrive_dashboard() {
...
+                                        <tr>
+                                            <?php
+                                            printf("<td>%s</td>", $row['table']);
+                                            printf("<td>%d</td>", $row['size']);
+                                            ?>
```
Comment (fuegas) [inline src/class-savvii-dashboard.php:224]: As you're already mixing HTML and PHP, why do you print the HTML?
```suggestion
                                            <td><?php echo $row['table']; ?></td>
                                            <td><?php echo $row['size']; ?></td>
```

**Pull request closed without merging**